### PR TITLE
Drop guava

### PR DESCRIPTION
**Pull request by dmandalidis**

### CODE REVIEW

File: src/main/java/org/mandas/docker/client/DefaultDockerClient.java [hunk 1]
```diff
@@ -969,7 +967,7 @@ public TopResults topContainer(final String containerId, final String psArgs)
       throws DockerException, InterruptedException {
     try {
       WebTarget resource = resource().path("containers").path(containerId).path("top");
-      if (!Strings.isNullOrEmpty(psArgs)) {
+      if (psArgs != null && "".equals(psArgs.trim())) {
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/DefaultDockerClient.java]: Shouldn't this be checking that psArgs is NOT empty?

File: src/main/java/org/mandas/docker/client/DefaultDockerClient.java [hunk 2]
```diff
@@ -2874,12 +2872,12 @@ public static Builder builder() {
...
-            DockerHost.configPathFromEnv(),
-            DockerHost.defaultCertPath()),
-        Predicates.notNull()));
-
+    final Path dockerCertPath = Paths.get(asList(certPathFromEnv(), configPathFromEnv(), defaultCertPath())
+    	.stream()
+    	.filter(cert -> cert != null)
+    	.findFirst()
+    	.orElseThrow(() -> new NoSuchElementException("Cannot find docker certificated path")));
+
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/DefaultDockerClient.java]: nit: whitespace

File: src/main/java/org/mandas/docker/client/DefaultDockerClient.java [hunk 3]
```diff
@@ -2874,12 +2872,12 @@ public static Builder builder() {
...
-        Arrays.asList(DockerHost.certPathFromEnv(),
-            DockerHost.configPathFromEnv(),
-            DockerHost.defaultCertPath()),
-        Predicates.notNull()));
-
+    final Path dockerCertPath = Paths.get(asList(certPathFromEnv(), configPathFromEnv(), defaultCertPath())
+    	.stream()
+    	.filter(cert -> cert != null)
+    	.findFirst()
+    	.orElseThrow(() -> new NoSuchElementException("Cannot find docker certificated path")));
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/DefaultDockerClient.java:2866]: I'm not very savvy with stream syntax, but this looks mostly reasonable.  The orElseThrow makes sense, but how does the Exception get handled when it occurs?

File: src/main/java/org/mandas/docker/client/DefaultLogStream.java [hunk 4]
```diff
@@ -23,16 +23,14 @@
...
-import com.google.common.collect.AbstractIterator;
-import com.google.common.io.Closer;
-
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.nio.ByteBuffer;
 
+import com.google.common.collect.AbstractIterator;
+import com.google.common.io.Closer;
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/DefaultLogStream.java]: Is this still a work in progress?  If not, were these missed?

File: src/main/java/org/mandas/docker/client/DockerHost.java [hunk 5]
```diff
@@ -23,14 +23,13 @@
...
-import com.google.common.base.MoreObjects;
-import com.google.common.net.HostAndPort;
-
 import java.net.URI;
 import java.nio.file.Paths;
 import java.util.Locale;
 
+import com.google.common.base.MoreObjects;
+import com.google.common.net.HostAndPort;
+
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/DockerHost.java]: More guava remnants

**Comment by tony--** [comment]
I like this idea - it's something my team has talked about but never explored.  I scanned through the changes quickly and it looks like this is still a work in progress?

File: src/main/java/org/mandas/docker/client/DefaultDockerClient.java (hunk 1 above)
Comment (dmandalidis) [reply src/main/java/org/mandas/docker/client/DefaultDockerClient.java]: good catch

File: src/main/java/org/mandas/docker/client/DefaultDockerClient.java (hunk 3 above)
Comment (dmandalidis) [reply src/main/java/org/mandas/docker/client/DefaultDockerClient.java:2866]: The same way as before; the replaced `Iterables.find` throws a `NoSuchElementException` if the element is not found

**Comment by dmandalidis** [comment]
Hi @tony-- 

I 'm done with this. Please ping me if you 're going to review it so I can wait a little bit before merging it

File: src/main/java/org/mandas/docker/client/DefaultLogStream.java (hunk 4 above)
Comment (dmandalidis) [reply src/main/java/org/mandas/docker/client/DefaultLogStream.java]: done

**Comment by tony--** [comment]
@dmandalidis - Thanks for the effort on this.  I'll take a look ASAP, but I might not get to it until Tuesday (US West Coast, +10 hours your time, I think)

**Comment by dmandalidis** [comment]
no problem, take yr time

File: CHANGELOG.md [hunk 6]
```diff
@@ -7,6 +7,12 @@
 * `RegistryConfigs.Builder#addConfig` does not ignore null `RegistryAuth` values anymore
 * `HostConfig.Builder#appendBinds*` methods removed
 * `HostConfig.Builder#binds(final Bind... binds)` method does not ignore null binds anymore
+* `Immutable*` types removed from public getters and builders
+* `EventStream` now implements `java.util.Iterator<Event>`
+* `EventStream#close` and `LogStream#close` can now throw `IOException`
+* `LogStream#attach(OutputStream, OutputStream, boolean)` method removed. Methods should
+not close streams they do not manage.
+* `LogStream#attach(OutputStream, OutputStream)` can now throw `IOException`
```
Comment (tony--) [inline CHANGELOG.md:15]: Consider adding a bullet for "removed dependency on Guava, which resulted in some of the breaking changes"

File: src/main/java/org/mandas/docker/client/AnsiProgressHandler.java [hunk 7]
```diff
@@ -46,7 +45,7 @@ public AnsiProgressHandler() {
 
   public AnsiProgressHandler(PrintStream out) {
     this.out = out;
-    idsToLines = Maps.newHashMap();
+    idsToLines = new HashMap<>();
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/AnsiProgressHandler.java:48]: ~~seems like this should be `idsToLines = new HashMap<String, Integer>();` in order to avoid a warning?~~

Update: I tried it in Eclipse and see that an untyped constructor is ok.

File: src/main/java/org/mandas/docker/client/CompressedDirectory.java [hunk 8]
```diff
@@ -137,14 +136,14 @@ public void close() throws IOException {
     Files.delete(file);
   }
 
-  static ImmutableList<DockerIgnorePathMatcher> parseDockerIgnore(Path dockerIgnorePath)
+  static List<DockerIgnorePathMatcher> parseDockerIgnore(Path dockerIgnorePath)
       throws IOException {
-    final ImmutableList.Builder<DockerIgnorePathMatcher> matchersBuilder = ImmutableList.builder();
+    final List<DockerIgnorePathMatcher> matchersBuilder = new ArrayList<>();
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/CompressedDirectory.java:142]: ~~Should `new ArrayList<>` be typed as `DockerIgnorePathMatcher` in order to avoid a warning?~~

Update: I tried it in Eclipse and see that an untyped constructor is ok.

File: src/main/java/org/mandas/docker/client/CompressedDirectory.java [hunk 9]
```diff
@@ -253,15 +251,16 @@ private static String getNotSeparatorPattern(String separator) {
...
     private final TarArchiveOutputStream tarStream;
 
-    private Visitor(final Path root, ImmutableList<DockerIgnorePathMatcher> ignoreMatchers,
+    private Visitor(final Path root, List<DockerIgnorePathMatcher> ignoreMatchers,
                     final TarArchiveOutputStream tarStream) {
       this.root = root;
       // .dockerignore matchers need to be read from the bottom of the file, 
       // so the given list should be reversed before using it.
-      this.ignoreMatchers = ignoreMatchers.reverse();
+      this.ignoreMatchers = new ArrayList<>(ignoreMatchers);
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/CompressedDirectory.java]: ~~should new ArrayList<> be typed?~~

Update: I tried it in Eclipse and see that an untyped constructor is ok.

Comment (tony--) [reply src/main/java/org/mandas/docker/client/CompressedDirectory.java]: For consistency, should this be an unmodifiableList?

File: src/main/java/org/mandas/docker/client/DefaultDockerClient.java [hunk 10]
```diff
@@ -264,7 +259,7 @@ private LoadProgressHandler(ProgressHandler delegate) {
     }
 
     private Set<String> getImageNames() {
-      return ImmutableSet.copyOf(imageNames);
+      return unmodifiableSet(new HashSet<>(imageNames));
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/DefaultDockerClient.java:262]: ~~Should this be typed?~~
~~I'm afraid all of these "should this be typed?" comments are a waste, but I'll keep calling out in case you want to type them.~~

Update: I tried it in Eclipse and see that an untyped constructor is ok.

File: src/main/java/org/mandas/docker/client/DefaultDockerClient.java [hunk 11]
```diff
@@ -431,8 +427,9 @@ public DefaultDockerClient(final URI uri, final DockerCertificatesStore dockerCe
    * @param builder DefaultDockerClient builder
    */
   protected DefaultDockerClient(final Builder builder) {
-    final URI originalUri = checkNotNull(builder.uri, "uri");
-    checkNotNull(originalUri.getScheme(), "url has null scheme");
+	requireNonNull(builder.uri, "uri");
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/DefaultDockerClient.java]: nit: indenting

File: src/main/java/org/mandas/docker/client/DefaultDockerClient.java [hunk 12]
```diff
@@ -536,7 +534,7 @@ private String toRegExp(String hostnameWithWildcards) {
 
   @Override
   public String getHost() {
-    return fromNullable(uri.getHost()).or("localhost");
+    return ofNullable(uri.getHost()).orElse("localhost");
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/DefaultDockerClient.java:537]: I hadn't used this before and had to read the docs.  Seems pretty cool!  Let's see if I understand.
if `uri.getHost()` is non-null, `ofNullable` returns an Optional containing the value of uri.getHost().
if `uri.getHost()` is null, `ofNullable` returns an empty Optional.
if the invoking Optional has a value, `orElse` returns that value.
if the invoking Optional does not have a value, `orElse` returns "localhost".
Is that right?

File: src/main/java/org/mandas/docker/client/DefaultDockerClient.java [hunk 13]
```diff
@@ -2737,7 +2737,7 @@ private void requestAndTail(final String method, final ProgressHandler handler,
     Response response = request(method, Response.class, resource, request);
     if (response.getStatusInfo().getFamily() != Response.Status.Family.SUCCESSFUL) {
       throw new DockerRequestException(method, resource.getUri(), response.getStatus(),
-                message(response), null);
+    		  response.readEntity(String.class), null);
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/DefaultDockerClient.java:2740]: Seems like a nice change here.

File: src/main/java/org/mandas/docker/client/DefaultDockerClient.java (hunk 2 above)
Comment (tony--) [reply src/main/java/org/mandas/docker/client/DefaultDockerClient.java]: This was marked resolved but it looks like the whitespace is still there?

File: src/main/java/org/mandas/docker/client/DefaultDockerClient.java [hunk 14]
```diff
@@ -2874,12 +2859,12 @@ public static Builder builder() {
...
-            DockerHost.configPathFromEnv(),
-            DockerHost.defaultCertPath()),
-        Predicates.notNull()));
-
+    final Path dockerCertPath = Paths.get(asList(certPathFromEnv(), configPathFromEnv(), defaultCertPath())
+    	.stream()
+    	.filter(cert -> cert != null)
+    	.findFirst()
+    	.orElseThrow(() -> new NoSuchElementException("Cannot find docker certificated path")));
+
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/DefaultDockerClient.java]: This was marked resolved but it looks like the whitespace is still there?

File: src/main/java/org/mandas/docker/client/DockerHost.java [hunk 15]
```diff
@@ -81,12 +72,17 @@ private DockerHost(final String endpoint, final String certPath) {
       this.bindUri = URI.create(endpoint);
     } else {
       final String stripped = endpoint.replaceAll(".*://", "");
-      final HostAndPort hostAndPort = HostAndPort.fromString(stripped);
-      final String hostText = hostAndPort.getHost();
-      final String scheme = isNullOrEmpty(certPath) ? "http" : "https";
-
-      this.port = hostAndPort.getPortOrDefault(defaultPort());
-      this.address = isNullOrEmpty(hostText) ? DEFAULT_ADDRESS : hostText;
+      final String scheme = certPath == null || "".equals(certPath.trim()) ? "http" : "https";
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/DockerHost.java]: It might be possible for certPath to be non-null but not exist.  Would it make sense to check if certPath.exists() and has certs in it?  I'm not sure, just thought of it after comparing with the similar code [here](https://github.com/dmandalidis/docker-client/pull/78/files#diff-497f14a9cf87565d6db0dab39d4d5a47R2879)

File: src/main/java/org/mandas/docker/client/DockerHost.java [hunk 16]
```diff
@@ -243,14 +242,14 @@ static String configPathFromEnv() {
...
-        .add("port", port)
-        .add("certPath", certPath)
-        .toString();
+	return "DockerHost{"
+		+ "host=" + host
+		+ ",uri=" + uri
+		+ ",bindUri=" + bindUri
+		+ ",address=" + address
+		+ ",port=" + port
+		+ ",certPath=" + certPath == null? "null": certPath
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/DockerHost.java]: if you decide to check for existence above, you should check for existence here also

File: src/main/java/org/mandas/docker/client/EventStream.java [hunk 17]
```diff
@@ -20,48 +20,51 @@
...
+  public boolean hasNext() {
+	if (next != null) {
+		return true;
+	}
+	try {
+		next = reader.nextMessage();
+	} catch (IOException e) {
+		throw new RuntimeException(e);
+	}
+	return next != null;
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/EventStream.java]: nit: indent is 4 instead of 2

File: src/main/java/org/mandas/docker/client/EventStream.java [hunk 18]
```diff
@@ -20,48 +20,51 @@
...
-    	throwIfUnchecked(e);
-        throw new RuntimeException(e);
-    }
+  public Event next() {
+	if (!hasNext()) {
+		throw new NoSuchElementException();
+	}
+	Event value = next;
+	next = null;
+	return value;
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/EventStream.java:68]: nit: indent is 4 instead of 2

File: src/main/java/org/mandas/docker/client/ImageRef.java [hunk 19]
```diff
@@ -87,15 +84,14 @@ public String getRegistryUrl() {
...
-    return MoreObjects.toStringHelper(this)
-        .add("registry", registry)
-        .add("image", image)
-        .add("tag", tag)
-        .toString();
+	return "ImageRef{"
+		+ "registry=" + registry != null? registry: null
+		+ ",image=" + image != null? image: null
+		+ ",tag=" + tag != null? tag: null
+		+ "}";
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/ImageRef.java]: nit: indent is 4 instead of 2

File: src/main/java/org/mandas/docker/client/LogMessage.java [hunk 20]
```diff
@@ -34,8 +34,10 @@ public LogMessage(final int streamId, final ByteBuffer content) {
   }
 
   public LogMessage(final Stream stream, final ByteBuffer content) {
-    this.stream = checkNotNull(stream, "stream");
-    this.content = checkNotNull(content, "content");
+	requireNonNull(stream, "stream");
+	requireNonNull(content, "content");
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/LogMessage.java:38]: nit: indent is 4 instead of 2

File: src/main/java/org/mandas/docker/client/LogReader.java [hunk 21]
```diff
@@ -66,15 +63,31 @@ public LogMessage nextMessage() throws IOException {
...
+    while (total < buf.length) {
+      int result = stream.read(buf, total, buf.length - total);
+      if (result == -1) {
+        break;
+      }
+      total += result;
+    }
+    
+    return total;
+  }
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/LogReader.java:87]: nit: indent is 4 instead of 2

File: src/main/java/org/mandas/docker/client/LogReader.java [hunk 22]
```diff
@@ -66,15 +63,31 @@ public LogMessage nextMessage() throws IOException {
...
+  
   @Override
   public void close() throws IOException {
-    // Jersey will close the stream and release the connection after we read all the data.
-    // We cannot call the stream's close method because it an instance of UncloseableInputStream,
-    // where close is a no-op.
-    copy(stream, nullOutputStream());
+	stream.close();
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/LogReader.java:91]: nit: indent is 4 instead of 2

File: src/main/java/org/mandas/docker/client/LogReader.java [hunk 23]
```diff
@@ -66,15 +63,31 @@ public LogMessage nextMessage() throws IOException {
...
+    if (total < frame.length) {
+      throw new EOFException("EOF before reading " + frame.length + " bytes. " + total + " read instead.");
+    }
+    
     return new LogMessage(streamId, ByteBuffer.wrap(frame));
   }
 
+  private int readAllBytes(byte[] buf) throws IOException {
+	int total = 0;
+    while (total < buf.length) {
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/LogReader.java:78]: Do you need this while loop and the 3 parameter version of `read`?  Reading the description [here](https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#read(byte[])) it seems like this will be functionally equivalent to `stream.read(buf);`

File: src/main/java/org/mandas/docker/client/auth/gcr/ContainerRegistryAuthSupplier.java [hunk 24]
```diff
@@ -127,7 +131,7 @@ public static Builder forCredentials(final GoogleCredentials credentials) {
     private final GoogleCredentials credentials;
 
     private Collection<String> scopes =
-        ImmutableList.of("https://www.googleapis.com/auth/devstorage.read_write");
+        Collections.singletonList("https://www.googleapis.com/auth/devstorage.read_write");
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/auth/gcr/ContainerRegistryAuthSupplier.java:134]: Nice, I had no idea this existed.

File: src/main/java/org/mandas/docker/client/messages/HostConfig.java [hunk 25]
```diff
@@ -314,8 +313,9 @@ public static Builder builder() {
   default void check() {
     if (extraHosts() != null) {
 	    for (final String extraHost : extraHosts()) {
-	      checkArgument(extraHost.contains(":"),
-	      "extra host arg '%s' must contain a ':'", extraHost);
+	      if (!extraHost.contains(".")) {
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/messages/HostConfig.java]: "." should be ":" here, I think?

File: src/main/java/org/mandas/docker/client/messages/RegistryConfigs.java [hunk 26]
```diff
@@ -66,22 +68,18 @@ public static RegistryConfigs create(final Map<String, RegistryAuth> configs) {
...
+    		RegistryAuth value = entry.getValue();
+    		if (value == null) {
+    			return null;
+    		}
+    		if (value.serverAddress() == null) {
+    			return value.toBuilder().serverAddress(entry.getKey()).build();
+    		}
+    		return value;
+    	}));
+
```
Comment (tony--) [inline src/main/java/org/mandas/docker/client/messages/RegistryConfigs.java:82]: nit: mix of 2, 4, 8 indent lengths

File: src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java [hunk 27]
```diff
@@ -1139,7 +1139,7 @@ public void testExportContainer() throws Exception {
     final ContainerCreation creation = sut.createContainer(config, name);
     final String id = creation.id();
 
-    final ImmutableSet.Builder<String> files = ImmutableSet.builder();
+    final Set<String> files = new HashSet<>();
```
Comment (tony--) [inline src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java]: nit: making this an unmodifiableSet would be more consistent

File: src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java [hunk 28]
```diff
@@ -783,7 +783,7 @@ public void testRemoveImage() throws Exception {
     final String imageLatest = "dxia/cirros:latest";
     final String imageVersion = "dxia/cirros:0.3.0";
 
-    final Set<RemovedImage> removedImages = Sets.newHashSet();
+    final Set<RemovedImage> removedImages = new HashSet<>();
```
Comment (tony--) [inline src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java]: nit: making this an unmodifiableSet would be more consistent

File: src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java [hunk 29]
```diff
@@ -1164,7 +1164,7 @@ public void testArchiveContainer() throws Exception {
     final ContainerCreation creation = sut.createContainer(config, name);
     final String id = creation.id();
 
-    final ImmutableSet.Builder<String> files = ImmutableSet.builder();
+    final Set<String> files = new HashSet<>();
```
Comment (tony--) [inline src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java]: nit: making this an unmodifiableSet would be more consistent

File: src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java [hunk 30]
```diff
@@ -1437,7 +1437,7 @@ public void integrationTest() throws Exception {
     }
 
     // Copy the same files from container
-    final ImmutableSet.Builder<String> filesDownloaded = ImmutableSet.builder();
+    final Set<String> filesDownloaded = new HashSet<>();
```
Comment (tony--) [inline src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java]: nit: making this an unmodifiableSet would be more consistent

File: src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java [hunk 31]
```diff
@@ -2367,7 +2367,7 @@ public void testListImages() throws Exception {
     // Can list by name
     final List<Image> imagesByName = sut.listImages(byName(BUSYBOX));
     assertThat(imagesByName.size(), greaterThan(0));
-    final Set<String> repoTags = Sets.newHashSet();
+    final Set<String> repoTags = new HashSet<>();
```
Comment (tony--) [inline src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java]: nit: making this an unmodifiableSet would be more consistent

File: src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java [hunk 32]
```diff
@@ -3476,9 +3476,9 @@ public void testListContainers() throws Exception {
   public void testContainerLabels() throws Exception {
     sut.pull(BUSYBOX_LATEST);
 
-    final Map<String, String> labels = ImmutableMap.of(
-        "name", "starship", "foo", "bar"
-    );
+    Map<String, String> labels = new HashMap<>();
```
Comment (tony--) [inline src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java]: nit: making this an unmodifiableSet would be more consistent

File: src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java [hunk 33]
```diff
@@ -3496,9 +3496,9 @@ public void testContainerLabels() throws Exception {
     final ContainerInfo containerInfo = sut.inspectContainer(id);
     assertThat(containerInfo.config().labels(), is(labels));
 
-    final Map<String, String> labels2 = ImmutableMap.of(
-        "name", "starship", "foo", "baz"
-    );
+    final Map<String, String> labels2 = new HashMap<>();
```
Comment (tony--) [inline src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java]: nit: making this an unmodifiableSet would be more consistent

File: src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java [hunk 34]
```diff
@@ -3637,8 +3637,9 @@ public void testNetworks() throws Exception {
         .driver("default")
         .config(singletonList(ipamConfig))
         .build();
-    final Map<String, String> labels = ImmutableMap.of(
-        "name", "starship", "foo", "bar");
+    final Map<String, String> labels = new HashMap<>();
```
Comment (tony--) [inline src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java]: nit: making this an unmodifiableSet would be more consistent

File: src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java [hunk 35]
```diff
@@ -4632,7 +4633,9 @@ public void testSecretOperations() throws Exception {
 
     final String secretData = Base64.getEncoder().encodeToString("testdata".getBytes(StandardCharsets.UTF_8));
     
-    final Map<String, String> labels = ImmutableMap.of("foo", "bar", "1", "a");
+    final Map<String, String> labels = new HashMap<>();
```
Comment (tony--) [inline src/test/java/org/mandas/docker/client/DefaultDockerClientTest.java]: nit: making this an unmodifiableSet would be more consistent

File: src/test/java/org/mandas/docker/client/DefaultLogStreamTest.java [hunk 36]
```diff
@@ -45,12 +45,13 @@ public void testAttach() throws Exception {
         null
     );
 
-    final ByteArrayOutputStream stdout = new ByteArrayOutputStream();
-    final ByteArrayOutputStream stderr = new ByteArrayOutputStream();
-    logStream.attach(stdout, stderr);
+    try (final ByteArrayOutputStream stdout = new ByteArrayOutputStream();
+    		final ByteArrayOutputStream stderr = new ByteArrayOutputStream()) {
+    	logStream.attach(stdout, stderr);
```
Comment (tony--) [inline src/test/java/org/mandas/docker/client/DefaultLogStreamTest.java:50]: nit: extra indent

Why did you need to add the try-with-resources here?

**Review by tony-- — CHANGES_REQUESTED** [review]
I'm done reviewing, please take a look at my comments.  Thanks for all the effort here.

File: src/main/java/org/mandas/docker/client/DefaultDockerClient.java (hunk 12 above)
Comment (dmandalidis) [reply src/main/java/org/mandas/docker/client/DefaultDockerClient.java:537]: correct

File: CHANGELOG.md (hunk 6 above)
Comment (dmandalidis) [reply CHANGELOG.md:15]: This would be added along with the commits just before releasing 3.0.0. I 've added these bullets here so I won't forget them :)

File: src/main/java/org/mandas/docker/client/DockerHost.java (hunk 15 above)
Comment (dmandalidis) [reply src/main/java/org/mandas/docker/client/DockerHost.java]: The `DockerHost#from(String, String)` is publicly exposed so technically the `certPath` can be anything. 

In general, I really didn't question the existing codebase to avoid mixing additional refactorings which would make this PR unreviewable. My gut feeling is that we 're checking too much stuff and we 're doing extensive checks just for protecting the clients sending us nonsense. I don't remember where, but I saw it multiple times resulting in correlation problems between what you 've passed and the error we print after sanitization. We could do the check you propose and fail early but I believe this is not in scope for this PR as well as addressing the code duplication between this method and the similar one in `DefaultDockerClient`

File: src/main/java/org/mandas/docker/client/messages/HostConfig.java (hunk 25 above)
Comment (dmandalidis) [reply src/main/java/org/mandas/docker/client/messages/HostConfig.java]: good catch

File: src/test/java/org/mandas/docker/client/DefaultLogStreamTest.java (hunk 36 above)
Comment (dmandalidis) [reply src/test/java/org/mandas/docker/client/DefaultLogStreamTest.java:50]: Didn't resist the temptation to close the byte array streams (resource leaks even in tests can cause problems as well)

File: src/main/java/org/mandas/docker/client/LogMessage.java (hunk 20 above)
Comment (dmandalidis) [reply src/main/java/org/mandas/docker/client/LogMessage.java:38]: it's not (seems like a github UI thing)

File: src/main/java/org/mandas/docker/client/LogReader.java (hunk 22 above)
Comment (dmandalidis) [reply src/main/java/org/mandas/docker/client/LogReader.java:91]: same github UI thing

File: src/main/java/org/mandas/docker/client/LogReader.java (hunk 23 above)
Comment (dmandalidis) [reply src/main/java/org/mandas/docker/client/LogReader.java:78]: That was my initial impression as well but tests failed at the end. The difference between the `stream.read(buf)` and this while loop is that the first one could return if just a single byte is read. What we need at this point is to block until we reach the buffer size or EOF. The previous `ByteStreams.readFully` was doing this internally as well

**Comment by dmandalidis** [comment]
@tony-- thanks a million for the effort reviewing this. I think I 've addressed your comments, please take a final look when you feel so.

As a side note, I think I 've been misusing the `Resolve conversation` button, I believe you should do that if you see that each one of your comments is addressed. Me doing this seems a bit rude, sorry :)

File: src/main/java/org/mandas/docker/client/DockerHost.java (hunk 15 above)
Comment (tony--) [reply src/main/java/org/mandas/docker/client/DockerHost.java]: Ok, sounds good, thanks for for explaining.

**Review by tony-- — APPROVED** [review]
LGTM
